### DISCUSSION TITLE
Best way to remove enclosed grains (grains fully within another grain)

### DISCUSSION THREAD

**Discussion by Remsdel** [mtex-toolbox/mtex#791] [Ask Anything] [answered]
Has anybody an efficient way to remove grain (and its boundary) that are fully enclosed within another grain?
The only way I could think so far is to search in grains.boundary.grainId for grains whose boundary is shared with only one other grains but I haven't worked out a working script yet. Eventually pixels belonging to the enclosed grains should be assigned to the surrounding grain. If somebody can give the few lines of code that it would require, it would be great.

**Comment by lfgmorales** [comment]
Have you look here:

https://mtex-toolbox.github.io/grain2d.checkInside.html

Cheers

Luiz

Luiz F. G. Morales
ScopeM - ETH Zürich
Sent from iPhone

On 19 Feb 2021, at 09:38, Remsdel <notifications@github.com> wrote:

﻿

Has anybody an efficient way to remove grain (and its boundary) that are fully enclosed within another grain?
The only way I could think so far is to search in grains.boundary.grainId for grains whose boundary is shared with only one other grains but I haven't worked out a working script yet. Eventually pixels belonging to the enclosed grains should be assigned to the surrounding grain. If somebody can give the few lines of code that it would require, it would be great.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/mtex-toolbox/mtex/discussions/791>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ALE5INZEYNP2J5BVVTO46TDS7YPRPANCNFSM4X36TDQQ>.

[ { "@context": "http://schema.org", "@type": "EmailMessage", "potentialAction": { "@type": "ViewAction", "target": "https://github.com/mtex-toolbox/mtex/discussions/791", "url": "https://github.com/mtex-toolbox/mtex/discussions/791", "name": "View Discussion" }, "description": "View this Discussion on GitHub", "publisher": { "@type": "Organization", "name": "GitHub", "url": "https://github.com" } } ]

**Reply by kilir** [reply]
And here:

https://mtex-toolbox.github.io/grain2d.isInclusion.html

Cheers,
Rüdiger

**Comment by Remsdel** [comment]
Thanks, that is indeed straightforward.
Once you have identified the grains, how would you remove them and make them part of the surrounding grain? 
My first take is to do something like  grains(selected_grains_id)=[ ], however I notice that this does not remove the grain boundary belonging to the deleted grains.

**Reply by kilir** [reply]
Like with any other list in  Matlab, `list_of_smothing(condition)=[]`. Here's an example https://mtex-toolbox.github.io/SelectingGrains.html#20

The same can be done for the ebsd, so you could reconstruct grains to incorporate the holes `ebsd(grains(condition)) = []`


Cheers,
Rüdiger

**Comment by nyyssont** [comment, marked as answer]
It looks like you want:

[grainsMerged,parentId] = merge(grains,'inclusions')

See:

https://mtex-toolbox.github.io/grain2d.merge.html

**Comment by Remsdel** [comment]
This is it, thank you

**Comment by Remsdel** [comment]
[Grains_failed_merge.zip](https://github.com/mtex-toolbox/mtex/files/6014973/Grains_failed_merge.zip)

The merge function appears to fail in some EBSD dataset:

[grainsMerged,parentId] = merge(grains,'inclusions');
Unable to perform assignment because the left and right sides have a different number of elements.

Error in grain2d/merge (line 255)
    grainsMerged.phaseId(i) = find(newPhase);


It works fine on other EBSD data belonging to the same batch of measurements, acquired the same way, on the same instrument and processed the same way in MTEX so the failure on some dataset must be caused by some limitations of the merge function. This is beyond my ability to dig out the issue so any help is welcome. Grains data is attached

**Reply by ralfHielscher** [reply]
The error appears because there is a grain which is entirely surrounded by a notIndexed grain.
The question is how to handle this in future. 

1. Shall we consider this grain as inclusion?
2. Shall this grain be merged into the surrounding not indexed grain?

Ralf.

**Comment by Remsdel** [comment]
I visualized some of the grains causing the issue. In all cases, they are indeed surrounded by notIndexed grain (due to porosity) and very small (just above the cleaning threshold limit) and very few (1 or 2 per map). As far I am concerned, I would have them merged with the surrounding not indexed grains (which after all is what the function is supposed to do). Would an update of the merge function be possible to handle these rare occurrences?
Thx

**Reply by ralfHielscher** [reply]
The file https://github.com/mtex-toolbox/mtex/blob/develop/EBSDAnalysis/%40grain2d/merge.m
contains the fix.

Ralf.

**Comment by Remsdel** [comment]
Great, thanks a lot.
Remi
